### ISSUE TITLE
Unhandled exception

### ISSUE THREAD

**Issue by Dalesw2** [woodenphone/Derpibooru-dl#33]
Keep getting this error on a few download requests.

2018-03-05 19:29:31,773 - DEBUG - https://derpibooru.org:443 "GET /search.json?perpage=50&page=11&q=artist%3Alumineko&key=11111111111111111111&filter_id=&sf=created_at&sd=desc HTTP/1.1" 200 None
2018-03-05 19:29:31,835 - DEBUG - Starting new HTTPS connection (1): derpibooru.org
2018-03-05 19:29:32,164 - DEBUG - https://derpibooru.org:443 "GET /search.json?perpage=50&page=12&q=artist%3Alumineko&key=11111111111111111111&filter_id=&sf=created_at&sd=desc HTTP/1.1" 200 None
2018-03-05 19:29:32,210 - CRITICAL - Unhandled exception!
2018-03-05 19:29:32,210 - CRITICAL - <type 'exceptions.AttributeError'>
2018-03-05 19:29:32,210 - ERROR - can't set attribute
Traceback (most recent call last):
  File "C:\Users\Dale\Desktop\Derpibooru-dl-master\derpibooru_dl.py", line 1411, in <module>
    main()
  File "C:\Users\Dale\Desktop\Derpibooru-dl-master\derpibooru_dl.py", line 1394, in main
    console_menu(settings,input_file_list)
  File "C:\Users\Dale\Desktop\Derpibooru-dl-master\derpibooru_dl.py", line 1309, in console_menu
    process_query(settings,search_query)
  File "C:\Users\Dale\Desktop\Derpibooru-dl-master\derpibooru_dl.py", line 987, in process_query
    submission_ids = search_for_query(settings, search_query)
  File "C:\Users\Dale\Desktop\Derpibooru-dl-master\derpibooru_dl.py", line 589, in search_for_query
    for image in derpibooru.Search().key(settings.api_key).limit(None).query(search_query):
  File "C:\Python27\lib\site-packages\derpibooru\search.py", line 197, in next
    return Image(self._search.next())
  File "C:\Python27\lib\site-packages\derpibooru\image.py", line 46, in __init__
    setattr(self, field, body)
AttributeError: can't set attribute
2018-03-05 19:29:32,226 - INFO - Program finished.

**Comment by Apple240Bloom** [comment]
Try running it with short file names on there were a couple I needed to do
that with because the file names were literally too large for the os.

**Comment by Dalesw2** [comment]
Tried dosen't work

**Comment by Dalesw2** [comment]
also this seem to be only happening on the pages listing part, not downloading

**Comment by Bendito999** [comment]
I can confirm this happens on mine as well on my saved queries (worked about a week ago, but a couple days ago now it does this). I haven't figured out exactly why yet its doing this either.
Example if I copy your debug statement there to the browser https://derpibooru.org/search.json?perpage=50&page=12&q=artist%3Alumineko&key=11111111111111111111&filter_id=&sf=created_at&sd
the information that is returned looks reasonable, though I don't have previous data saved to compare with to see if the format changed; will require some more indepth troubleshooting I think (need to dust off my python debugging skills).

**Comment by woodenphone** [comment]
I'm sorry but I don't have time to really maintain this anymore because I'm much busier IRL now than I was when I wrote this downloader.

**Comment by Apple240Bloom** [comment]
Nah man it's all good, I hope you're doing well. Doesn't GitHub allow
people to continue your work? Sounds like there are some talented people
interested in keeping this train running, don't worry about it. You do you
bud.

**Comment by TwiDash34** [comment]
Друзья, программа сломалась? :( Кто ни будь может ее починить?

**Comment by Twi-Hard** [comment]
Could the people at Derpibooru maybe know what's changed? I've used the same install of this app without problems since 2016 (until now).

**Comment by Bendito999** [comment]
https://derpibooru.org/meta/derpibooru-dl-stopped-working-possible-change-in-how-api-results-are-returned 

Forum post created to see if the wonderful devs have any insight to what might have changed on it.

**Comment by Bendito999** [comment]
From forum: "Yes, it returns integer ids now instead of strings."
Well that's a good clue, I haven't figured out quite yet where to make that change to the script yet to properly handle that, but I'm thinking it might not be too hard, will likely work on it more this weekend if someone else doesn't fix it first.

**Comment by woodenphone** [comment]
Github should make it easy to "fork" this repository and from there you can publish to that as needed.
I've actually forgotten how most of this code works, but I'd suggest putting a breakpoint or print statement at the line of code it crashes at so you can see exactly what value is incorrect if you want to try fixing it.

"File "C:\Python27\lib\site-packages\derpibooru\image.py", line 46, in init
setattr(self, field, body)"
tells me that it's probably a derpibooru python library (Derpybooru or something like that, an interface between my code and derpibooru's web api IIRC) that I didn't actually write that's malfunctioning
At some point in derpibooru-dl's development I switched from handling everyhthing in my own code to using a library for it, if that library isn't working it may be needed to 'undo' the changes and handle everything directly again

It may be less work to build a whole new downloader than to make my one work nicely.
Feel free to reuse whatever code I wrote for this if you do.

TL;DR: Derpibooru-dl is a mess and you're welcome to use it to build something that works better, but I'm busy with real life.

**Comment by ImintA2** [comment]
Seems like there are some problems with the lib's properties mapping.
I edited line 53 of "lib\site-packages\derpibooru\image.py"
```
def tags(self):
    return self.data["tags"].split(", ")
```
to
```
  def tags(self):
    try:
        return self.data["tags"].split(", ")
    except:
        return []
```
and the script is working again for me.

Of course, this is just telling the script to ignore the exception and is not thoroughly tested, so try it at your own risk. 
I'm too lazy to find out the reason for the error since I just want it to work. :stuck_out_tongue:

Remember to recompile the script after editing if you're using the .exe version.

**Comment by Dalesw2** [comment]
Testing it now, if it's good I'll try to upload the exe.

**Comment by Dalesw2** [comment]
Tested on a small list of problem tags(32),all went through.
[derpibooru_dl.zip](https://github.com/woodenphone/Derpibooru-dl/files/1794640/derpibooru_dl.zip)
Might be buggy(still testing it) but so far seems ok.

Thank's everyone.

**Comment by Apple240Bloom** [comment]
You're awesome


On Mar 8, 2018 3:16 PM, "Dalesw2" <notifications@github.com> wrote:

Tested on small list of problem tags(32),all went through.
derpibooru_dl.zip
<https://github.com/woodenphone/Derpibooru-dl/files/1794640/derpibooru_dl.zip>

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/woodenphone/Derpibooru-dl/issues/33#issuecomment-371610848>,
or mute the thread

**Comment by Apple240Bloom** [comment]
You're awesome! =)

**Comment by liamwhite** [comment]
@woodenphone I am a [developer and administrator](https://derpibooru.org/profiles/byte%255B%255D) for Derpibooru.org. If you would like to transfer ownership of this project to me, I can maintain it.

**Comment by Bendito999** [comment]
Nice ImintA2 that fix worked for me as well!
Just to put in concrete example the difference between the formats as earlier described, it can be seen in the saved .json files for example
1st entry in new ones goes like
{"id":1670618
Old format is something like
{"id":"64575"
I'm thinking that the definition somewhere of the expected contents of that needs to change, but haven't quite found it as I'm not strong with the Python.
(or maybe i'm completely misunderstanding how its using json that's quite possible too)

**Comment by liamwhite** [comment]
This error should be fixed now. (`tags` was null for 41758 images).

**Comment by Bendito999** [comment]
Looks like that did it! It works without the try/except part in it now. Nice work!
